templates/user_profile.html

```diff
@@ -0,0 +1,377 @@
+                    $('#researcher_type_display').hide();
+                    $('#max_load_display').hide();
+                }
+            }
+
+            // Initial check on page load
+            toggleSupervisorSelect();
+
+            $('#is_researcher').change(function () {
+                toggleSupervisorSelect();
+            });
+        });
+    </script>
+    <script>
+        $(document).ready(function () {
+            $('#addPreferencesButton').click(function () {
+                $('#staticPreferencesTable').hide();
+                $('#editablePreferencesTable').show();
+                $('#addPreferencesButton').hide();
+                $('#savePreferencesButton').show();
+                $('#cancelEditButton').show();
+                $('#addRowButton').show();
+            });
+
+            $('#cancelEditButton').click(function () {
+                $('#editablePreferencesTable').hide();
+                $('#staticPreferencesTable').show();
+                $('#addPreferencesButton').show();
+                $('#savePreferencesButton').hide();
+                $('#cancelEditButton').hide();
+                $('#addRowButton').hide();
+            });
+
+            $('#addRowButton').click(function () {
+                const newRow = `
+                    <tr>
+                        <td>
+                            <select class="form-control course-select">
+                                {% for course in courses %}
+                                    <option value="{{ course.id }}-{{ course.year }}">
+                                        {{ course.code }} - {{ course.title }}
+                                    </option>
+                                {% endfor %}
+                            </select>
+                        </td>
+                        <td>
+                            <i class="fas fa-trash delete-icon"></i>
+                            <i class="fas fa-arrows-alt move-icon"></i>
+                        </td>
+                    </tr>
+                `;
+                $('#editablePreferencesTable tbody').append(newRow);
+                $("#editablePreferencesTable tbody").sortable({
+                    items: 'tr',
+                    handle: '.move-icon',
+                    axis: 'y',
+                    cursor: 'move'
+                });
+            });
+
+            $(document).on('click', '.delete-icon', function () {
+                $(this).closest('tr').remove();
+            });
+
+            $('#savePreferencesButton').click(function () {
+                const preferences = [];
+                $('#editablePreferencesTable tr').each(function () {
+                    const courseCode = $(this).find('select.course-select').val();
+                    if (courseCode) {
+                        preferences.push(courseCode);
+                    }
+                });
+
+                const data = {
+                    preferences: preferences
```

Comment:
Do you actually need to encapsulate your preferences into a one-key dict ?

Reply:
I used a dictionary because for each preference I have to save the year of the course and its id. Is there any other way?